src/api/common/connection.js

```diff
@@ -0,0 +1,125 @@
+'use strict';
+const {EventEmitter} = require('events');
+const WebSocket = require('ws');
+
+function createDisconnectionError(id) {
+  return {id: id, status: 'error', type: 'error', error: 'disconnected'};
+}
+
+function isStreamMessageType(type) {
+  return type === 'ledgerClosed' ||
+         type === 'transaction' ||
+         type === 'path_find';
+}
+
+class Connection extends EventEmitter {
+  constructor(url) {
+    super();
+    this._url = url;
+    this._ws = null;
+    this._nextRequestID = 1;
+  }
+
+  _onMessage(message) {
+    try {
+      const data = JSON.parse(message);
+      if (data.type === 'response') {
+        if (!(Number.isInteger(data.id) && data.id >= 0)) {
+          throw new Error('valid id not found in response');
+        }
+        this.emit(data.id.toString(), data);
+      } else if (isStreamMessageType(data.type)) {
+        this.emit(data.type, data);
+      } else {
+        throw new Error('unrecognized message type: ' + data.type);
```

Comment:
`slowDown` error will be emitted as `badMessage`, without meaningful description.